x/stakeibc/keeper/unbonding_records.go

```diff
@@ -235,6 +245,93 @@ func (k Keeper) GetUnbondingICAMessages(
+	unbondingsCapped := initialUnbondings[:batchSize]
+
+	// Calculate the amount that was initially meant to be unbonded from that batch,
+	// and determine the remainder that needs to be redistributed
+	unbondAmountFromBatch := sdkmath.ZeroInt()
+	initialUnbondingsMap := map[string]sdkmath.Int{}
+	for _, unbonding := range unbondingsCapped {
+		unbondAmountFromBatch = unbondAmountFromBatch.Add(unbonding.Amount)
+		initialUnbondingsMap[unbonding.Validator] = unbonding.Amount
+	}
+	totalExcessAmount := totalUnbondAmount.Sub(unbondAmountFromBatch)
+
+	// Store the delegation of each validator that was expected *after* the originally
+	// planned unbonding went through
+	// e.g. If the validator had 10 before unbonding, and in the first pass, 3 was
+	//      supposed to be unbonded, their delegation after the first pass is 7
+	totalDelegationsAfterFirstPass := sdk.ZeroDec()
+	delegationsAfterFirstPass := map[string]sdk.Dec{}
+	for _, capacity := range unbondCapacities {
+		// Only add validators that were in the initial unbonding plan
+		// The delegation after the first pass is calculated by taking the "current delegation"
+		// (aka delegation before unbonding) and subtracting the unbond amount
+		if initialUnbondAmount, ok := initialUnbondingsMap[capacity.ValidatorAddress]; ok {
+			delegationAfterFirstPass := sdk.NewDecFromInt(capacity.CurrentDelegation.Sub(initialUnbondAmount))
+
+			delegationsAfterFirstPass[capacity.ValidatorAddress] = delegationAfterFirstPass
+			totalDelegationsAfterFirstPass = totalDelegationsAfterFirstPass.Add(delegationAfterFirstPass)
+		}
+	}
+
+	// This should never happen, but it protects against a division by zero error below
+	if totalDelegationsAfterFirstPass.IsZero() {
+		return finalUnbondings, errors.New("no delegations to redistribute during consolidation")
+	}
+
+	// Loop through the original unbonding messages and proportionally divide out
+	// the excess amongst the validators in the set
+	excessRemaining := totalExcessAmount
+	for i := range unbondingsCapped {
+		unbonding := unbondingsCapped[i]
+
+		var validatorUnbondIncrease sdkmath.Int
+		if i != len(unbondingsCapped)-1 {
+			// For all but the last validator, calculate their unbonding increase by
+			// splitting the excess proportionally in line with their remaining delegation
+			delegationAfterFirstPass, ok := delegationsAfterFirstPass[unbonding.Validator]
+			if !ok {
+				return finalUnbondings, fmt.Errorf("validator not found in initial unbonding plan")
+			}
+			unbondIncreaseProportion := delegationAfterFirstPass.Quo(totalDelegationsAfterFirstPass)
+			validatorUnbondIncrease = sdk.NewDecFromInt(totalExcessAmount).Mul(unbondIncreaseProportion).TruncateInt()
+
+			// Decrement excess
+			excessRemaining = excessRemaining.Sub(validatorUnbondIncrease)
+		} else {
+			// The last validator in the set should get any remainder from int truction
+			validatorUnbondIncrease = excessRemaining
```

Comment:
especially here we should check that validatorUnbondIncrease <= delegationAfterFirstPass on this last validator -- maybe we should even sort the capped set by validatorUnbondIncrease (increasing) so this final catch all would fall onto the validator with the most extra delegation in the set?

Reply:
I think [this](https://github.com/Stride-Labs/stride/pull/1018/commits/c28539c1b1056747e2d1806e715b551864236c10#diff-ba71e9ef6bf1ebcc5243683cbd44405947d8109d7ffb825c7c353d256f7e2179R297-R301) check is sufficient, but added another check [here](https://github.com/Stride-Labs/stride/pull/1018/commits/e388120788759f76b792e4e1988a98f6623bc705) since it's free and gives a bit more peace of mind